File: src/org/cerion/symcalc/expression/ConstExpr.java
package org.cerion.symcalc.expression;

import org.cerion.symcalc.expression.constant.Pi;

import java.util.Hashtable;

public abstract class ConstExpr extends Expr {

	public static boolean isConstant(String s) {
		return (lookup(s) != null);
	}

	public static ConstExpr getConstant(String name) {
		Name n = lookup(name);
		switch(n) {
			case PI: return new Pi();
		}

		throw new IllegalArgumentException("invalid constant");
	}

	@Override
	public ExprType getType() {
		return ExprType.CONST;
	}

	@Override
	public abstract  String toString();

	public void show(int i) {
        indent(i,"Constant: " + toString());
	}

	@Override
	public boolean equals(Expr e) {
		return getClass() == e.getClass();
	}

	private enum Name {
		PI,
		E
	}

	private static Hashtable<String, Name> identifiers;
	static 
    {
    	identifiers = new Hashtable<>();
    	identifiers.put("pi", Name.PI);
		identifiers.put("e", Name.E);
    }

	private static Name lookup(String s) {
		String lookup = s.toLowerCase();
		if(identifiers.containsKey(lookup))
			return identifiers.get(lookup);

		return null;
	}

	@Override
	protected int getProperties() {
		return Properties.CONSTANT.value;
	}
}

File: src/org/cerion/symcalc/expression/ConstExpr.kt
package org.cerion.symcalc.expression

import org.cerion.symcalc.expression.constant.Pi

import java.util.Hashtable

abstract class ConstExpr : Expr() {

    override fun getType(): Expr.ExprType = Expr.ExprType.CONST
    override fun show(i: Int) = indent(i, "Constant: " + toString())
    override fun equals(e: Expr): Boolean = javaClass == e.javaClass
    override fun getProperties(): Int = Expr.Properties.CONSTANT.value

    abstract override fun toString(): String

    private enum class Name {
        PI,
        E
    }

    companion object {

        @JvmStatic fun isConstant(s: String): Boolean {
            return lookup(s) != null
        }

        @JvmStatic fun getConstant(name: String): ConstExpr {
            val n = lookup(name)
            when (n) {
                ConstExpr.Name.PI -> return Pi()
            }

            throw IllegalArgumentException("invalid constant")
        }

        private var identifiers: Hashtable<String, Name>? = null

        init {
            identifiers = Hashtable()
            identifiers!!["pi"] = Name.PI
            identifiers!!["e"] = Name.E
        }

        private fun lookup(s: String): Name? {
            val lookup = s.toLowerCase()
            return if (identifiers!!.containsKey(lookup)) identifiers!![lookup] else null

        }
    }
}
